src/log/dmlog_appender.cpp

```diff
@@ -0,0 +1,65 @@
+#include <fc/log/dmlog_appender.hpp>
+#include <fc/log/log_message.hpp>
+#include <fc/string.hpp>
+#include <fc/variant.hpp>
+#include <fc/reflect/variant.hpp>
+#ifndef WIN32
+#include <unistd.h>
+#endif
+#include <boost/asio/io_context.hpp>
+#include <boost/thread/mutex.hpp>
+#include <fc/exception/exception.hpp>
+#include <iomanip>
+#include <mutex>
+#include <sstream>
+
+namespace fc {
+   class dmlog_appender::impl {
+      public:
+         boost::asio::io_service* io_service;
+         boost::mutex log_mutex;
+   };
+
+   dmlog_appender::dmlog_appender( const variant& args )
+   :my(new impl){}
+
+   dmlog_appender::dmlog_appender()
+   :my(new impl){}
+
+   dmlog_appender::~dmlog_appender() {}
+
+   void dmlog_appender::initialize( boost::asio::io_service& io_service ) {
+      my->io_service = &io_service;
+   }
+
+   void dmlog_appender::log( const log_message& m ) {
+      FILE* out = stdout;
+
+      string message = format_string( "DMLOG " + m.get_format() + "\n", m.get_data() );
+      std::unique_lock<boost::mutex> lock(my->log_mutex);
+
+      int retries = 0;
+      auto remaining_size = message.size();
+      auto message_ptr = message.c_str();
+      while (true) {
+         auto written = fwrite(message_ptr, sizeof(char), remaining_size, out);
+         if (written == remaining_size) {
+            break;
+         }
+
+         fprintf(stderr, "DMLOG FPRINTF_FAILED failed written=%lu remaining=%lu %d %s\n", written, remaining_size, ferror(out), strerror(errno));
+
+         if (retries++ > 5) {
```

Comment:
anything else needs to do when retry limits reached?

Reply:
Not really, it's time to exit and check the machine, there is something bad happening.

I however added a `is_stopped` state to the logger since the `io_service->stop` is not in effect immediately. 

Should be enough like that.